doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,259 @@
+    - UNIX epoch as common reference point for high interoperability
+    - Support for 64 bit time stamps
+        - Justification: Devices running for decades might want to log events
+          with milli second resolution timestamps on flash / sd card / ...
+        - Note: The rest of the API can be 32 bit, only for time stamps 64 bit
+          support is mandatory
+    - Optional feature to auto-initialize system time from RTC/RTT during boot
+    - APIs to allow implementing network time synchronization
+        - The synchronization mechanism should be highly accurate. (E.g. consider the mechanism would read the value from hardware, modify it by adding a time offset determined by some network time synchronization mechanism to it, and store the new value in the hardware: The actually applied offset would be lower by the time that has passed between the read from the hardware and the write to it, reducing accuracy. The actual implementation should be smarter.)
+        - Only the system time (the portable 64 bit time stamps) need to be synchronized.
+- Utilities for convenience
+    - Timer utilities
+        - Timer that send a message upon expiration
+        - Timer that post an event upon expiration (to allow execution in thread context)
+        - Timer that wakes up a thread upon expiration
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+    - System time utilities
+        - Functions to convert between different time formats, e.g. UNIX and GPS timestamps
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time. More
+  specifically: With *n* software timers used in parallel, the longest period of
+  time with interrupts disabled be at most *O(n)*
+- The implementation must aim for a minimized number of IRQs caused
+- The implementation must aim for keeping its ISRs short; with *n* software
+  timers used in parallel, the ISR should run for at most *O(n)*
+- The implementation should provide means to execute callbacks in thread
+  context
+
+Justification: Real time capable systems require an upper bound for the latency
+a high level timer implementation provides. For most application - if not all -
+it is trivial to get an upper bound for the number of software timers being
+active in parallel. Hence, an *O(f(n))* latency with *n* being the number of
+active software timers makes the real time behavior of a system predictable and
+easy to estimate. In addition, *f(n)* should increase as slowly as possible,
+while not being to strict to rule out reasonable trade-offs e.g. between
+ROM/RAM requirements and latency. An *O(n)* latency is e.g. feasible to
+implement with a sorted linked list and therefore a reasonable lower bar.
+
```

Comment:
*O(n)* is an unbound function, so this justification is inconsistent. For real-time requirements ask for *O(1)*, I suppose.

Reply:
For every concrete application, there is an upper bound for the number *n* of software timers being active at the same time. So for every concrete application, *O(n)* has an upper bound.

Reply:
This is true for any finite set of numbers. So we could also go for an *O(exp(n))*?

Reply:
> _O(n)_ is an unbound function, so this justification is inconsistent. For real-time requirements ask for _O(1)_, I suppose.

This is intentionally written to _not_ be O(1), because of the reasonable trade-offs as described. We know how to do O(n) by simply using linked lists and sorting on insert, and we understand the runtime implications well. We don't know how to do this in O(1) without limiting the number of allowed timers.

We do multiplexing here, so one hardware timer (or more, but a small number) needs to be used to schedule N timeouts.
So when adding one of these timeouts (using a set(interval) function), that new timeout needs to be added into the list of already queued timers. Using linked lists, this is an O(n) operation. Using an array and binary search, this is O(logn). We need to be careful what to require here, because if we break O(n*logn) for sorting n timers, we might attract a lot of attention.

It is possible to add an allowed timers limit (that is lower than available memory, e.g., 16 or 128) and enforce this.
We'd suddenly have O(1) on paper (meeting the requirements), but wouldn't have gained anything, other than that now a multitude of functions have an error case (timer_sleep() might fail with "maximum number of timers reached") that needs to be checked everywhere, which is unreasonable.

TL;DR let's not require something we don't know how to implement. This is phrased as "yadda yadda reasonable trade-offs".

Reply:
The complete timer management can be done in O(log(n)), which is almost constant in the number range we consider realistically. Further optimizations can apply, if algorithms account for the deadline of the next firing timer, which can be determined in constant time. 
Current preliminary HiL-tests of @pokgak show a rather poor behavior of xtimer and a somewhat unclear signature of ztimer. We will publish once ready. 
In any case, it is important that timer firing remains within predictable bounds independent of the number of instantiated timers - otherwise real-time gets out of control.

Reply:
> The complete timer management can be done in O(log(n)), which is almost constant in the number range we consider realistically.

I thought we all agree on the fact that *O(f(n))* is constant for any *f(n)* within the number range we consider realistically?

Reply:
Off course, we don't.

Reply:
> _O(n)_ is an unbound function, so this justification is inconsistent. For real-time requirements ask for _O(1)_, I suppose.

Nope, soft and hard real-time system only care if the growth is bounded at some point, they require a *the worst case* execution time (in units of time or instructions).

The growth function doesn't provide any information on how long the function takes to execute. An _O(1)_ function is just as bad as an n-limited O(n) function if it exceeds the permitted execution time, the only given is that the O(1) will always exceed the deadline and that the O(n) might exceed the deadline in some cases.

As long as the developer can have a guarantee that the number of simultaneous active timers is bounded, the worst case execution time can be determined and whether the system will function can be reasoned about. Doesn't matter here if the growth function is O(1), O(n), O(log(n)) or O(exp(n)) as those do not tell us anything about the worst case execution time, only how it scales if the number of active timers is increased.

Reply:
> Doesn't matter here if the growth function is O(1), O(n), O(log(n)) or O(exp(n)) as those do not tell us anything about the worst case execution time, only how it scales if the number of active timers is increased.

???

Reply:
> ???

Care to elaborate what exactly is not clear here?

Reply:
log(25) ~ 1.39,  exp(25) ~ 7.2 x 10**10.

Reply:
> log(25) ~ 1.39, exp(25) ~ 7.2 x 10**10.

Both bounded, which is exactly what a realtime system needs.

Reply:
Really?

Reply:
> Really?

Do you want to dive into the numerical analysis of the numbers you provided or can we agree here that both numbers are constants?

A realtime system does not require the system to be fast. It does however require the system to respond within the specified deadlines. This deadline can be in the order of microsecond or in the order of years, depending on the requirements.

Furthermore, a complex O(1) algorithm, taking 10 seconds performs worse compared with a simple O(n) algorithm taking `10 ms + n * 1 ms`. when n is known to be bounded to 100.

Reply:
As a solution here, why don't we rephrase a bit and specify that it must be trivial to determine the maximum number of active timers for an application in order to ensure bounded execution time?